src/compiler/parser.ts

```diff
@@ -2313,6 +2345,9 @@ module ts {
         }
 
         function parseTypeWorker(): TypeNode {
+            if (isStartOfTypeGuardType()) {

```

Comment:
This will do a look ahead for almost every type node. I would not do it this way. Instead, I'd have a function called parseTypeGuardTypeOrHigher, that does something like this:
```ts
let type = parseUnionTypeOrHigher();
while (token === SyntaxKind.IsKeyword) {
    // create a new TypeGuardTypeNode 'node'
    node.target = type;
    // more stuff
    type = node;
}
return type;
```
Basically what parseUnionTypeOrHigher does, but for type guard types instead.

I am assuming here that you want type guard types to be the loosest precedence, so that
```ts
x is Foo | Bar
```
means that the parameter x is guarded to be of type `Foo | Bar`.